File: services/api/src/api/routers/system.py
"""System endpoints: health, queues, workers, metrics (PRD §9)."""

from __future__ import annotations

from typing import Any

import httpx
from core.config import get_settings
from core.observability.metrics import metrics
from core.queue import streams
from fastapi import APIRouter, Depends
from redis import Redis
from sqlalchemy import text
from sqlalchemy.orm import Session

from api.deps import get_session

router = APIRouter(prefix="/v1/system", tags=["system"])


def _check_postgres(session: Session) -> str:
    try:
        session.execute(text("SELECT 1"))
        return "ok"
    except Exception:
        return "down"


def _check_redis() -> str:
    try:
        streams.make_redis().ping()
        return "ok"
    except Exception:
        return "down"


def _check_url(url: str) -> str:
    try:
        return (
            "ok" if httpx.get(f"{url.rstrip('/')}/health", timeout=2).status_code == 200 else "down"
        )
    except Exception:
        return "down"


def _bearer(key: str | None) -> dict[str, str]:
    return {"Authorization": f"Bearer {key}"} if key else {}


def _check_qdrant(url: str) -> str:
    """Qdrant has no /health — probe /collections root with the API key."""
    try:
        s = get_settings()
        r = httpx.get(
            url.rstrip("/") + "/collections",
            timeout=2,
            headers={"Authorization": f"Bearer {s.qdrant_api_key}"} if s.qdrant_api_key else {},
        )
        return "ok" if r.status_code == 200 else "down"
    except Exception:
        return "down"


@router.get("/health")
def health(session: Session = Depends(get_session)):
    s = get_settings()
    return {
        "status": "ok" if _check_postgres(session) == "ok" else "degraded",
        "postgres": _check_postgres(session),
        "redis": _check_redis(),
        "qdrant": _check_qdrant(
            s.qdrant_url
        ),  # /collections probe WITH api key (/health 404s + needs auth)
        "tei_query": _check_url(s.tei_query_url),
    }


@router.get("/queues")
def queues():
    """Depth + consumer lag per stream (PRD §10.1).

    undelivered comes from streams.undelivered_count() — NULL-safe lag with
    a live-count fallback (2026-09-12: XINFO lag=None after XDEL cleanup
    collapsed to 0 here, and the main dashboard disagreed with the pipeline
    page). Errors propagate: a dead Redis must not read as an empty queue.
    """
    r: Redis = streams.make_redis()
    out = {}
    for name in streams.ALL_STREAMS:
        try:
            pending = r.xpending(name, streams.CONSUMER_GROUP)
            out[name] = {
                # "length" is XLEN (every entry ever written — streams are
                # never trimmed), NOT the live backlog. Kept for history.
                "length": r.xlen(name),
                "pending": pending["pending"] if pending else 0,
                "undelivered": streams.undelivered_count(r, name),
            }
        except Exception:
            raise
    return out


@router.get("/pipeline")
def pipeline(session: Session = Depends(get_session)):
    """Aggregate snapshot for the pipeline live view (web /pipeline page).

    One call answers: is every component up, how deep is each queue lane,
    who is holding in-flight work, and how fast are shards settling.
    """
    s = get_settings()
    r: Redis = streams.make_redis()

    # -- component health (reuse /health checks + ollama ingest probe) ------
    components = {
        "postgres": _check_postgres(session),
        "redis": _check_redis(),
        "qdrant": _check_qdrant(s.qdrant_url),
        "tei_query": _check_url(s.tei_query_url),
    }
    # embed backend: ollama (Jetson) answers /api/tags, TEI answers /health
    try:
        components["embed_backend"] = (
            "ok"
            if httpx.get(f"{s.tei_ingest_url.rstrip('/')}/api/tags", timeout=2).status_code == 200
            else "down"
        )
    except Exception:
        components["embed_backend"] = "down"

    # -- queue lanes: waiting / in-flight / stale / consumers ---------------
    lanes = {}
    for name in streams.ALL_STREAMS:
        lane = {"waiting": None, "in_flight": None, "stale": 0, "consumers": 0}
        try:
            # waiting = undelivered + PEL, NOT XLEN — streams are never
            # trimmed so XLEN only ever grows (it showed 20,703 "waiting"
            # on a queue with 0 undelivered + 23 pending).
            # undelivered_count() handles Redis lag=None (untrusted
            # bookkeeping, e.g. after XDEL) with a live count fallback —
            # int(grp.get("lag") or 0) collapsed that to 0 and hid a
            # 73-job backlog on 2026-09-12.
            grp = next(
                (g for g in r.xinfo_groups(name) if g.get("name") == streams.CONSUMER_GROUP),
                None,
            )
            undelivered = (
                streams.undelivered_count(r, name)
                if grp is not None
                else 0
            )
            pend = r.xpending_range(name, streams.CONSUMER_GROUP, min="-", max="+", count=100)
            lane["waiting"] = undelivered + len(pend)
            lane["in_flight"] = len(pend)
            now_ms = r.time()[0]
            for entry in pend:
                if now_ms - int(entry["time_since_delivered"]) > 900_000:  # 15 min
                    lane["stale"] += 1
            consumers = r.xinfo_consumers(name, streams.CONSUMER_GROUP)
            lane["consumers"] = sum(1 for c in consumers if c.get("idle", 10**9) < 300_000)
        except Exception:
            pass
        lanes[name] = lane

    # -- document/shard counters --------------------------------------------
    counts: dict[str, Any] = {}
    try:
        row = (
            session.execute(
                text(
                    "SELECT"
                    "  count(*) FILTER (WHERE state='ready') AS ready,"
                    "  count(*) FILTER (WHERE state='parsing') AS parsing,"
                    "  count(*) FILTER (WHERE state='failed') AS failed,"
                    "  count(*) AS total FROM documents"
                )
            )
            .mappings()
            .one()
        )
        counts["docs_ready"] = row["ready"]
        counts["docs_parsing"] = row["parsing"]
        counts["docs_failed"] = row["failed"]
        counts["docs_total"] = row["total"]

        # parsing breakdown (derived, 2026-09-12): state='parsing' hides two
        # phases — shards still converting vs whole-book settled waiting for
        # the embedder (whole-book barrier). Break them out so the dashboard
        # can show parser vs embedder lag separately. No new enum value —
        # derived from shards counters (PG is the source of truth).
        brow = (
            session.execute(
                text(
                    "SELECT"
                    "  count(*) FILTER (WHERE total_shards IS NOT NULL AND shards_done >= total_shards) AS awaiting_embed,"
                    "  count(*) FILTER (WHERE total_shards IS NULL OR shards_done < total_shards) AS parsing_active"
                    " FROM documents WHERE state='parsing'"
                )
            )
            .mappings()
            .one()
        )
        counts["docs_awaiting_embed"] = brow["awaiting_embed"]
        counts["docs_parsing_active"] = brow["parsing_active"]

        srow = (
            session.execute(
                text(
                    "SELECT"
                    "  count(*) FILTER (WHERE state='done') AS done,"
                    "  count(*) FILTER (WHERE state='pending') AS pending,"
                    "  count(*) FILTER (WHERE state='failed') AS failed,"
                    "  count(*) FILTER (WHERE state='running') AS running,"
                    "  count(*) AS total FROM shards"
                )
            )
            .mappings()
            .one()
        )
        counts["shards_done"] = srow["done"]
        counts["shards_pending"] = srow["pending"]
        counts["shards_failed"] = srow["failed"]
        counts["shards_running"] = srow["running"]
        counts["shards_total"] = srow["total"]

        # Full state breakdown — dashboards must NOT derive per-state counts
        # from ?limit=N document lists (the /dashboard "ready" card read the
        # first 200 rows and undercounted vs the pipeline aggregate forever).
        strow = (
            session.execute(text("SELECT state, count(*) AS n FROM documents GROUP BY state"))
            .mappings()
            .all()
        )
        counts["docs_states"] = {r["state"]: r["n"] for r in strow}
    except Exception:
        pass

    # -- in-flight parse detail: shards with a live lease --------------------
    in_flight = []
    try:
        rows = session.execute(
            text(
                "SELECT d.title, s.idx, s.page_start, s.page_end,"
                " s.worker_id, s.lease_until"
                " FROM shards s JOIN documents d ON d.id = s.doc_id"
                " WHERE s.state='running' LIMIT 12"
            )
        )
        in_flight = [dict(x) for x in rows.mappings()]
    except Exception:
        pass

    # -- qdrant points (chunks collection) -----------------------------------
    qdrant_points = None
    try:
        qr = httpx.get(
            s.qdrant_url.rstrip("/") + "/collections/chunks",
            timeout=3,
            headers={"Authorization": f"Bearer {s.qdrant_api_key}"} if s.qdrant_api_key else {},
        )
        if qr.status_code == 200:
            qdrant_points = qr.json().get("result", {}).get("points_count")
    except Exception:
        pass

    return {
        "components": components,
        "lanes": lanes,
        "counts": counts,
        "in_flight_parse": in_flight,
        "qdrant_points": qdrant_points,
        "server_time": None,  # web layer stamps local time
    }


@router.get("/metrics")
def get_metrics():
    """JSON counters, Prometheus-convertible shape (§10.4)."""
    return metrics().snapshot()

File: internal/api/system.go
package api

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/saufi-opi/lybrix/internal/queue"
	"github.com/saufi-opi/lybrix/internal/store"
)

// healthPayload mirrors SystemHealth semantics: status ok|degraded, plus
// per-component strings. Qdrant's slot is gone; the embed backend probe
// (tei-ingest / ollama) is exposed at /pipeline, not here (system.py kept
// the health body to postgres/redis/qdrant/tei_query — 2.0 swaps qdrant
// for nothing at this level; the store is Postgres itself).
type healthPayload struct {
	Status   string `json:"status"`
	Postgres string `json:"postgres"`
	Redis    string `json:"redis"`
	TeiQuery string `json:"tei_query"`
}

// checkURL probes {url}/health with a 2s timeout; "ok" only on HTTP 200.
func checkURL(ctx context.Context, url string) string {
	if url == "" {
		return "down"
	}
	client := &http.Client{Timeout: 2 * time.Second}
	resp, err := client.Get(url + "/health")
	if err != nil {
		return "down"
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		return "ok"
	}
	return "down"
}

func (s *Server) handleHealth(w http.ResponseWriter, r *http.Request) {
	pg := "ok"
	if err := s.deps.DB.Ping(r.Context()); err != nil {
		pg = "down"
	}
	redisStatus := "ok"
	if err := s.deps.Redis.Ping(r.Context()).Err(); err != nil {
		redisStatus = "down"
	}
	tei := "down"
	if def, err := s.deps.DB.GetDefaultEmbeddingModel(r.Context()); err == nil && def != nil {
		tei = checkURL(r.Context(), def.QueryURL)
	}
	status := "ok"
	if pg != "ok" {
		status = "degraded"
	}
	WriteJSON(w, http.StatusOK, healthPayload{
		Status: status, Postgres: pg, Redis: redisStatus, TeiQuery: tei,
	})
}

func (s *Server) handleQueues(w http.ResponseWriter, r *http.Request) {
	out := map[string]any{}
	for _, name := range queue.AllStreams {
		length, err := s.deps.Redis.XLen(r.Context(), name).Result()
		if err != nil {
			// a dead Redis must not read as an empty queue — propagate (500)
			writeDetail(w, http.StatusInternalServerError, err.Error())
			return
		}
		pending, err := s.deps.Redis.XPending(r.Context(), name, queue.ConsumerGroup).Result()
		if err != nil {
			pending = nil
		}
		var pcount int64
		if pending != nil {
			pcount = pending.Count
		}
		undelivered, err := queue.UndeliveredCount(r.Context(), s.deps.Redis, name)
		if err != nil {
			writeDetail(w, http.StatusInternalServerError, err.Error())
			return
		}
		out[name] = map[string]any{
			// "length" is XLEN (every entry ever written — streams are never
			// trimmed), NOT the live backlog. Kept for history.
			"length":      length,
			"pending":     pcount,
			"undelivered": undelivered,
		}
	}
	WriteJSON(w, http.StatusOK, out)
}

// laneOut is one stream's lane view (pipeline page).
type laneOut struct {
	Waiting   *int   `json:"waiting"`
	InFlight  *int   `json:"in_flight"`
	Stale     int    `json:"stale"`
	Consumers int    `json:"consumers"`
	Error     string `json:"error,omitempty"`
}

func (s *Server) handlePipeline(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// -- component health -------------------------------------------------
	components := map[string]any{
		"postgres": "ok",
		"redis":    "ok",
	}
	if err := s.deps.DB.Ping(ctx); err != nil {
		components["postgres"] = "down"
	}
	if err := s.deps.Redis.Ping(ctx).Err(); err != nil {
		components["redis"] = "down"
	}
	components["tei_query"] = "down"
	components["embed_backend"] = "down"
	components["embedding_model"] = ""
	components["embed_provider"] = ""
	if def, err := s.deps.DB.GetDefaultEmbeddingModel(ctx); err == nil && def != nil {
		components["tei_query"] = checkURL(ctx, def.QueryURL)
		components["embed_backend"] = checkEmbedModel(ctx, def)
		components["embedding_model"] = def.ModelID
		components["embed_provider"] = def.Provider
	}

	// -- queue lanes -------------------------------------------------------
	lanes := map[string]any{}
	for _, name := range queue.AllStreams {
		lane := laneOut{}
		groups, err := s.deps.Redis.XInfoGroups(ctx, name).Result()
		if err != nil {
			slog.Warn("pipeline lane read failed", "stream", name, "err", err)
			lane.Error = err.Error()
			lanes[name] = lane
			continue
		}
		var grp *queueGroupInfo
		for _, g := range groups {
			if g.Name == queue.ConsumerGroup {
				grp = &queueGroupInfo{LastDelivered: g.LastDeliveredID}
				break
			}
		}
		if grp == nil {
			lanes[name] = lane
			continue
		}
		undelivered, err := queue.UndeliveredCount(ctx, s.deps.Redis, name)
		if err != nil {
			lane.Error = err.Error()
			lanes[name] = lane
			continue
		}
		pend, err := s.deps.Redis.XPendingExt(ctx, redisPendingExtArgs(name)).Result()
		if err != nil {
			lane.Error = err.Error()
			lanes[name] = lane
			continue
		}
		waiting := int(undelivered + int64(len(pend)))
		inFlight := len(pend)
		for _, p := range pend {
			// entries idle > 15 min are stale (system.py 900_000 ms check)
			if p.Idle > 15*time.Minute {
				lane.Stale++
			}
		}
		consumers, _ := s.deps.Redis.XInfoConsumers(ctx, name, queue.ConsumerGroup).Result()
		for _, c := range consumers {
			if c.Idle < 5*time.Minute {
				lane.Consumers++
			}
		}
		lane.Waiting = &waiting
		lane.InFlight = &inFlight
		lanes[name] = lane
	}

	// -- counters -----------------------------------------------------------
	counts := map[string]any{}
	if snap, err := s.deps.DB.MetricsSnapshot(ctx); err == nil {
		counts["docs_ready"] = snap.DocsReady
		counts["docs_parsing"] = snap.DocsParsing
		counts["docs_failed"] = snap.DocsFailed
		counts["docs_total"] = snap.DocsTotal
		counts["docs_awaiting_embed"] = snap.DocsAwaitingEmbed
		counts["docs_parsing_active"] = snap.DocsParsingActive
		counts["shards_done"] = snap.ShardsDone
		counts["shards_pending"] = snap.ShardsPending
		counts["shards_failed"] = snap.ShardsFailed
		counts["shards_running"] = snap.ShardsRunning
		counts["shards_total"] = snap.ShardsTotal
		// Full state breakdown — dashboards must NOT derive per-state counts
		// from ?limit=N document lists.
		counts["docs_states"] = snap.States
	}

	// -- in-flight parse detail ---------------------------------------------
	inFlight := []map[string]any{}
	if rows, err := s.deps.DB.InFlightParse(ctx); err == nil {
		for _, row := range rows {
			inFlight = append(inFlight, map[string]any{
				"title": row.Title, "idx": row.Idx,
				"page_start": row.PageStart, "page_end": row.PageEnd,
				"worker_id": row.WorkerID, "lease_until": row.LeaseUntil,
			})
		}
	}

	// -- chunk totals (2.0: ParadeDB replaces qdrant_points) -----------------
	chunksTotal := (*int)(nil)
	if n, err := s.deps.DB.CountChunks(ctx); err == nil {
		chunksTotal = &n
	}

	WriteJSON(w, http.StatusOK, map[string]any{
		"components":      components,
		"lanes":           lanes,
		"counts":          counts,
		"in_flight_parse": inFlight,
		// 2.0 delta: qdrant_points is replaced by chunks_total from ParadeDB.
		"chunks_total": chunksTotal,
		"server_time":  nil, // web layer stamps local time
	})
}

// queueGroupInfo carries the fields of XInfoGroups the lanes loop needs.
type queueGroupInfo struct {
	LastDelivered string
}

// checkEmbedModel probes one registered model's ingest-plane endpoint:
// ollama answers /api/tags, openai answers GET /v1/models, TEI /health.
func checkEmbedModel(ctx context.Context, m *store.EmbeddingModel) string {
	client := &http.Client{Timeout: 2 * time.Second}
	path := "/health"
	switch m.Provider {
	case "ollama":
		path = "/api/tags"
	case "openai":
		path = "/v1/models"
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, strings.TrimRight(m.IngestURL, "/")+path, nil)
	if err != nil {
		return "down"
	}
	if m.Provider == "openai" && m.HasAPIKey {
		// probe without the key: presence is enough for a reachability check;
		// the key itself must never transit logs/probe paths unnecessarily.
		_ = m
	}
	resp, err := client.Do(req)
	if err != nil {
		return "down"
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		return "ok"
	}
	return "down"
}

func (s *Server) handleMetrics(w http.ResponseWriter, r *http.Request) {
	// JSON counters snapshot: in-process atomic counters + janitor rollup
	// read (§10.4 Prometheus-convertible shape).
	snap := MetricsSnapshotJSON()
	rollup := map[string]any{}
	if rows, err := s.deps.DB.LatestRollups(r.Context(), 60); err == nil {
		for _, row := range rows {
			b, _ := json.Marshal(row)
			var m map[string]any
			_ = json.Unmarshal(b, &m)
			rollup[row.Bucket.Format(time.RFC3339)] = m
		}
	}
	snap["rollup"] = rollup
	WriteJSON(w, http.StatusOK, snap)
}

var _ = store.BookSettled

// redisPendingExtArgs builds the 1.0 pending-range shape: min "-" max "+",
// count 100.
func redisPendingExtArgs(stream string) *redis.XPendingExtArgs {
	return &redis.XPendingExtArgs{Stream: stream, Group: queue.ConsumerGroup,
		Start: "-", End: "+", Count: 100}
}
